kernel_gateway/services/kernels/handlers.py

```diff
@@ -51,6 +54,54 @@ def get_json_body(self):
                 model.setdefault('name', self.settings['kg_default_kernel_name'])
         return model
 
+class KernelHandler(TokenAuthorizationMixin,
+                    CORSMixin,
+                    notebook_handlers.KernelHandler):
+    '''This class will remove a kernel from the activity store when it is delted.
+    '''
+    @json_errors
+    def delete(self, kernel_id):
+        activity.remove(kernel_id)
+        super(KernelHandler, self).delete(kernel_id)
+
+class ZMQChannelsHandler(TokenAuthorizationMixin,
+                        notebook_handlers.ZMQChannelsHandler):
+    '''This class listens for messages coming to and from the kernel to provide metrics
+    about kernel usage.
+    '''
+    def open(self, kernel_id):
+        activity.publish(self.kernel_id, LAST_CLIENT_CONNECT, datetime.now().isoformat())
+        activity.increment_activity(self.kernel_id, CONNECTIONS)
+        super(ZMQChannelsHandler, self).open(kernel_id)
+
+    def on_close(self):
+        activity.publish(self.kernel_id, LAST_CLIENT_DISCONNECT, datetime.now().isoformat())
+        activity.decrement_activity(self.kernel_id, CONNECTIONS)
+        super(ZMQChannelsHandler, self).on_close()
+
+    def _on_zmq_reply(self, stream, msg_list):
+        msg_metadata = json.loads(msg_list[3].decode('UTF-8'))
+        # If the message coming back is a status message, we need to inspect it
+        if msg_metadata['msg_type'] == 'status':
+            msg_content = json.loads(msg_list[6].decode('UTF-8'))
```

Comment:
Parsing every message when the activity handler endpoints are not exposed will probably degredate performance. Can we add a quick if-check here and bypass all of this if the activity API is not enabled on startup?